removed the copy restriction when initializing a Table from a dict with copy=False

broulston opened the pull request:
removes the copy restriction when initializing a Table from a dict with copy=False from issue #8538 

I've simply removed the check on copy. I think there should be some updates to the documentation as well, but wanted to check before I went ahead and worked on that. I'm assuming I'll need to update the docs/table/construct_table.rst files, in the "Initialization Details" and "Copy versus Reference" sections?

bsipocz commented:
@broulston - Thanks for the PR. Please note the day regarding the bot's comment, hope it didn't cause confusion.

The ValueError in "Copy vs Reference" were raised in L299, and I couldn't spot which line you refer in "Initialization details". 
However it would be super useful to make sure there are tests that cover the cases that would have otherwise ended up in this conditional.

taldcroft commented:
@broulston - echoing @bsipocz, I'm a little dismayed (at myself) that there was no test that failed when you removed the line.  But indeed we want to fix that and add a test that proves that copy=False works with a dict and results in a reference (no copy) to the original data.

taldcroft commented:
And yes, in addition to any narrative doc updates you need to update the CHANGES.rst changelog as an astropy.table enhancement for 3.2.

I haven't looked in the narrative docs.  Probably the only change you need is if there are places that specifically mention this situation (init with dict and `copy=False`).  But you probably don't need to invent new examples or anything like that.

taldcroft commented:
@bsipocz - I see you put the `no-changelog` tag, but I think this does need a changelog.  I for one have been annoyed by this for a while (stupid I didn't just fix it) and would welcome seeing this in a changelog!

bsipocz commented:
@taldcroft - fair enough, I somehow assumed this is just a cleanup of a bigger PR that has covered it in the changelog already.

broulston commented:
@bsipocz, @taldcroft thanks for the feedback! I'll add to the changelog, and check the docs to see if there are any places that need an update. I'll start working on the tests for this as well. 

Would putting the test in "test_init_table.py" be correct?

Also, @taldcroft, you said to add to the changelog as an enhancement for 3.2 but I only see sections for "New Features", "API Changes", "Bug Fixes", and "Other Changes and Additions".

bsipocz commented:
@broulston - enhancements go under "new features".

taldcroft commented:
@broulston - BTW I just noticed you are here at CfA working with PG, nice!

broulston commented on astropy/table/tests/test_init_table.py:
```diff
@@ -478,3 +478,28 @@ def test_init_and_ref_from_multidim_ndarray(table_type):
+    Test that initializing from a dict works for both copy=False and True and that
+    the referencing is as expected.
+    """
+    for copy in (False, True):
```
@taldcroft ok, so by using "@pytest.mark.parametrize('copy', [False, True])", the loop isn't needed and pytest checks each of the values from the decorator?

bsipocz replied:
Basically yes. If there are multiple decorators then it checks all the combinations.

broulston commented on astropy/table/tests/test_init_table.py:
```diff
@@ -478,3 +478,28 @@ def test_init_and_ref_from_multidim_ndarray(table_type):
+    for copy in (False, True):
+        x1 = np.arange(10.)
+        x2 = np.zeros(10)
+        x3 = np.random.randint(10, size=10)
```
@taldcroft makes perfect sense! Removed

broulston commented:
It appears that the Travis CI had been stuck since yesterday. If you go to the details, it looks like build jobs 22880.3 and 22880.7 failed. 22880.7 doesn't look like it has to do with anything I changed, but 22880.3 looks like it failed because of the test I added. 

```
        col_dict = dict([('x1', x1), ('x2', x2)])
        t = table_type(col_dict, copy=copy)
>       assert t.colnames == ['x1', 'x2']
E       AssertionError: assert ['x2', 'x1'] == ['x1', 'x2']
E         At index 0 diff: 'x2' != 'x1'
E         Use -v to get the full diff
```
I don't quite get what has happened, since the tests run fine on my computer, and I think my logic is good in the test. It looks like the colnames list got reversed somehow.

bsipocz commented:
@broulston - That test run uses python 3.5 where dicts are not ordered by default, so getting those columns in a different order is explained.

I don't think we need to care here about the order of the columns (or the user should know to not use dicts in the first place), so I would suggest to just change the test.

bsipocz commented on astropy/table/tests/test_init_table.py:
```diff
@@ -478,3 +478,26 @@ def test_init_and_ref_from_multidim_ndarray(table_type):
+    x2 = np.zeros(10)
+    col_dict = dict([('x1', x1), ('x2', x2)])
+    t = table_type(col_dict, copy=copy)
+    assert t.colnames == ['x1', 'x2']
```
```suggestion
    assert set(t.colnames) == set(['x1', 'x2'])
```

broulston commented:
Thanks @bsipocz, I did not know that!

bsipocz commented:
@broulston - no worries, CI's job is to catch these.

broulston commented:
@taldcroft , @bsipocz I think this is ready for review
